### PR TITLE
Remove expensive trace events

### PR DESCRIPTION
**Pull request by jonahwilliams**
On applications that have a lot of composited layers, the cost of trace events can easily dwarf the actual raster time. See https://github.com/flutter/flutter/issues/113879#issuecomment-1289870517

This patch attempts to work around this by 1) removing trace events that provide low value or are called two frequently (display list) and 2) moving preroll/paint trace events to debug only.

### CODE REVIEW

File: display_list/display_list.cc [hunk 1]
```diff
@@ -75,7 +75,6 @@ void DisplayList::ComputeRTree() {
 void DisplayList::Dispatch(Dispatcher& dispatcher,
                            uint8_t* ptr,
                            uint8_t* end) const {
-  TRACE_EVENT0("flutter", "DisplayList::Dispatch");
```
Comment (dnfield) [inline display_list/display_list.cc:78]: Let's delete this and the one on `DisplayListDispatcher::EndRecordingAsPicture` in their own PR.

File: fml/trace_event.h [hunk 2]
```diff
@@ -70,6 +71,17 @@
   ::fml::tracing::TraceCounter((category_group), (name), (counter_id), (arg1), \
                                __VA_ARGS__);
 
+// Certain trace events happen so frequently that in complex apps the tracing
+// overhead can easily dwarf the actual raster time. For these, we use
+// a special debug mode only trace macro.
+#if (FLUTTER_RUNTIME_MODE_DEBUG)
+#define DEBUG_TRACE_EVENT0(category_group, name)     \
```
Comment (dnfield) [inline fml/trace_event.h:78]: It'd be nice if there was some way to turn these on or off via a shell argument, kind of like how we have `--trace-skia` to enable extra skia tracing today.

Having these in a trace can be helpful to understand scene complexity, but maybe there are better ways we can tool that (like @iskakaushik 's work around rasterizing leaves and timing that).

Comment (Hixie) [reply fml/trace_event.h:78]: The argument to turn this kind of thing on is `--profile`.

Comment (dnfield) [reply fml/trace_event.h:78]: Leaving these on by default in profile mode is likely artificially slowing down benchmarks, especially on iOS where tracing is more expensive.

I'm suggesting treating these more like how we treat Skia's helpful but too-verbose tracing in profile mode.

Comment (jonahwilliams) [reply fml/trace_event.h:78]: We could even leave them on by default, but we need to provide a mechanism to turn them off so our benchmarks are representative

**Comment by jonahwilliams** [comment]
Trying to summarize a lot of different discussion threads:

1. We don't want to add more tracing flags
2. We don't use these trace events for anything
3. These trace events aren't actually useful for debugging performance because of how expensive they are to record
4. Lots of discusson on whether or not we should trace in release on Android, but I don't really want to touch that :) 


I think that the best short term course of action is to remove these trace events. Worst comes to worst, we can add them back. But most of what is being traced here are very trivial operations that do not actually have much overhead at all (at least, not without the trace events running).

We already have tools for folks to debug what is in the layer tree.

**Review by zanderso — APPROVED** [review]

File: flow/layers/layer.cc [hunk 3]
```diff
@@ -77,7 +76,6 @@ Layer::AutoSaveLayer::AutoSaveLayer(const PaintContext& paint_context,
       canvas_(save_mode == SaveMode::kInternalNodesCanvas
                   ? *(paint_context.internal_nodes_canvas)
                   : *(paint_context.leaf_nodes_canvas)) {
-  TRACE_EVENT0("flutter", "Canvas::saveLayer");
```
Comment (kenzieschmoll) [inline flow/layers/layer.cc:80]: removing the saveLayer events will break DevTools. We have logic to count these events and provide a hint to users when we detect this expensive operation in their janky frame.

File: flow/layers/physical_shape_layer.cc [hunk 4]
```diff
@@ -103,7 +101,6 @@ void PhysicalShapeLayer::Paint(PaintContext& context) const {
       context.internal_nodes_canvas->clipPath(path_, true);
       break;
     case Clip::antiAliasWithSaveLayer: {
-      TRACE_EVENT0("flutter", "Canvas::saveLayer");
```
Comment (kenzieschmoll) [inline flow/layers/physical_shape_layer.cc:106]: same comment here

File: flow/layers/layer.cc (hunk 3 above)
Comment (jonahwilliams) [reply flow/layers/layer.cc:80]: So I can certainly leave these in the tree, but I don't think relying on this is correct. You are only going to see saveLayers that are composited by flutter, you will miss all of the saveLayers that are performed directly on cavnas. That could easily be the majority of cases

Comment (jonahwilliams) [reply flow/layers/layer.cc]: done

File: flow/layers/physical_shape_layer.cc (hunk 4 above)
Comment (jonahwilliams) [reply flow/layers/physical_shape_layer.cc]: done

File: flow/layers/layer.cc (hunk 3 above)
Comment (kenzieschmoll) [reply flow/layers/layer.cc]: Okay gotcha. This was the event I was told to check for at the time these were added - @dnfield what is the best way to catch all the save layer events from flutter and from the canvas?

Comment (jonahwilliams) [reply flow/layers/layer.cc]: Its not going to be possible without instrumenting the display list. that said, given that these trace events are slow we should find a different way to record this information.

**Comment by flar** [comment]
The overrides in all of the clip layers are now obsolete. They existed solely to send the event. I'll update this in the LayersStateStack PR. (Please do not create another PR to address this as I'll have to merge with it)

File: flow/layers/layer.cc (hunk 3 above)
Comment (dnfield) [reply flow/layers/layer.cc]: saveLayer opeations on canvas should not happen very often. Finding a different way to instrument this is fine, but we should make sure to not break devtools in the mean time.

Comment (flar) [reply flow/layers/layer.cc]: Thanks to this discussion I realized that I had not supported these events in the new LayerStateStack stuff.

https://github.com/flutter/engine/pull/36458/commits/7632996a3e0fa57db4b604f99a8939b7e53b9dc6

**Pull request merged**